src/lib/timing.c

```diff
@@ -39,53 +39,51 @@
 
 ZEND_EXTERN_MODULE_GLOBALS(xdebug)
 
-// Windows
-#if PHP_WIN32
 static uint64_t xdebug_get_nanotime_abs(xdebug_nanotime_context *nanotime_context)
 {
-	FILETIME filetime;
-
-	if (nanotime_context->win_precise_time_func != NULL) {
-		nanotime_context->win_precise_time_func(&filetime);
-		return ((((uint64_t)filetime.dwHighDateTime << 32) + (uint64_t)filetime.dwLowDateTime) - WIN_TICKS_SINCE_1601_JAN_1)
-			* WIN_NANOS_IN_TICK;
+#if PHP_WIN32
+	// Windows, win_precise_time_func is only available on Win 8 and later.
+	{
+		FILETIME filetime;
+
+		if (nanotime_context->win_precise_time_func != NULL) {
+			nanotime_context->win_precise_time_func(&filetime);
+			return ((((uint64_t)filetime.dwHighDateTime << 32) + (uint64_t)filetime.dwLowDateTime) - WIN_TICKS_SINCE_1601_JAN_1)
+				* WIN_NANOS_IN_TICK;
+		}
 	}
+#endif
 
-	return 0;
-}
-
-// Linux/Unix with clock_gettime
-#elif HAVE_CLOCK_GETTIME
-static uint64_t xdebug_get_nanotime_abs(xdebug_nanotime_context *nanotime_context)
-{
-   struct timespec ts;
+#if HAVE_CLOCK_GETTIME
+	// Linux/Unix with clock_gettime
+	{
+	   struct timespec ts;
 
 #ifdef CLOCK_MONOTONIC
-	if (clock_gettime(CLOCK_MONOTONIC, &ts) == 0) {
-		return (uint64_t)ts.tv_sec * NANOS_IN_SEC + (uint64_t)ts.tv_nsec;
-	}
-#else
-	if (clock_gettime(CLOCK_REALTIME, &ts) == 0) {
-		return (uint64_t)ts.tv_sec * NANOS_IN_SEC + (uint64_t)ts.tv_nsec;
+		if (clock_gettime(CLOCK_MONOTONIC, &ts) == 0) {
+			return (uint64_t)ts.tv_sec * NANOS_IN_SEC + (uint64_t)ts.tv_nsec;
+		}
+#endif
+		if (clock_gettime(CLOCK_REALTIME, &ts) == 0) {
+			return (uint64_t)ts.tv_sec * NANOS_IN_SEC + (uint64_t)ts.tv_nsec;
+		}
 	}
 #endif
 
-	return 0;
-}
+#if HAVE_GETTIMEOFDAY
+	// Fallback to gettimeofday() if better platform specific time is not available
+	{
+		struct timeval tp;
 
-// Fallback if better platform specific time is not available
-#elif HAVE_GETTIMEOFDAY
-static uint64_t xdebug_get_nanotime_abs(xdebug_nanotime_context *nanotime_context)
-{
-	struct timeval tp;
-
-	if (gettimeofday(&tp, NULL) == 0) {
-		return (uint64_t)tp.tv_sec * NANOS_IN_SEC + (uint64_t)tp.tv_usec * NANOS_IN_MICROSEC;
+		if (gettimeofday(&tp, NULL) == 0) {
+			return (uint64_t)tp.tv_sec * NANOS_IN_SEC + (uint64_t)tp.tv_usec * NANOS_IN_MICROSEC;
+		}
 	}
+#endif
 
+	// We give up
 	return 0;
```

Comment:
can we emit some warning to stderr or even exit (handle it like another critical handled erros in xdebug)? I think this should never be reached, `gettimeofday` should be always available fallback with non zero result.

Reply:
I've added an `E_WARNING` now. It turned out that although PHP on Windows does have `gettimeofday()`, it doesn't set `HAVE_GETTIMEOFDAY`, hence my change to the `ifdef` in line 73.

Reply:
`gettimeofday` was originally the primary and only time source and on Windows it worked (it was only inaccurate) - so test like originally